### PR TITLE
📖 update clusterctl provider contract for clusterclass support

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

This PR update the clusterctl provider contract for providers to opt into cluster class support.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: docs/book/src/clusterctl/provider-contract.md
Comment (sbueringer) [inline docs/book/src/clusterctl/provider-contract.md:259]:
```suggestion
1. The cluster class definition should be named `clusterclass-{ClusterClass-name}.yaml`, e.g `clusterclass-prod.yaml`
```
I think you can inline l.261 by slightly changing the var name

File: docs/book/src/clusterctl/provider-contract.md
Comment (sbueringer) [inline docs/book/src/clusterctl/provider-contract.md:269]:
```suggestion
All objects in the cluster class definition YAML MUST be deployed in the same namespace.
```
I wonder if we should recommend that the namespace is not set.

What would happen if someone sets a specific namespace? Would it be overwritten automatically or only when `${NAMESPACE}` is used?

File: docs/book/src/clusterctl/provider-contract.md
Comment (sbueringer) [inline docs/book/src/clusterctl/provider-contract.md:271]:
```suggestion
It is strongly recommended that none of the objects in the cluster class definitions are shared across multiple definitions.
```
nit

File: docs/book/src/clusterctl/provider-contract.md
Comment (sbueringer) [inline docs/book/src/clusterctl/provider-contract.md:280]:
```suggestion
if the cluster template uses a managed topology.
```
nit, not sure what fits better

File: docs/book/src/clusterctl/provider-contract.md
Comment (sbueringer) [inline docs/book/src/clusterctl/provider-contract.md:252]: Should we consistently use `ClusterClass` ?

**Review by sbueringer — COMMENTED** [review]
A few nits, but sounds good in general

File: docs/book/src/clusterctl/provider-contract.md
Comment (killianmuldoon) [inline docs/book/src/clusterctl/provider-contract.md:267]:
```suggestion
The cluster class definition YAML MUST assume the target namespace already exists.
```

File: docs/book/src/clusterctl/provider-contract.md
Comment (killianmuldoon) [inline docs/book/src/clusterctl/provider-contract.md:275]:
```suggestion
Currently the cluster class definition SHOULD NOT have any variables in it.
```
or 
```suggestion
Currently the cluster class definitions SHOULD NOT have any variables in them.
```

**Review by killianmuldoon — CHANGES_REQUESTED** [review]

File: docs/book/src/clusterctl/provider-contract.md
Comment (fabriziopandini) [inline docs/book/src/clusterctl/provider-contract.md:250]: nit: let's use consistently ClusterClass given that it is an API type
```suggestion
An infrastructure provider could publish a **ClusterClass definition** file to be used by `clusterctl generate cluster` that will be used along
```

File: docs/book/src/clusterctl/provider-contract.md
Comment (fabriziopandini) [reply docs/book/src/clusterctl/provider-contract.md:271]:
```suggestion
Even if technically possible, it is strongly recommended that none of the objects in the ClusterClass definitions are shared acorss multiple definitions; this helps on preventing that changing on an object impacts inadvertently many ClusterClasses, and consequently, all the Clusters using those.
```

File: docs/book/src/clusterctl/provider-contract.md
Comment (fabriziopandini) [inline docs/book/src/clusterctl/provider-contract.md:273]: This contrast with L263 ...

File: docs/book/src/clusterctl/provider-contract.md
Comment (fabriziopandini) [reply docs/book/src/clusterctl/provider-contract.md:280]:
```suggestion
if the workload cluster uses a managed topology and a ClusterClass with the same name does not already exists in the Cluster
```

**Review by fabriziopandini — COMMENTED** [review]
A couple of nits from my side
Q: should we update also documentation in https://cluster-api.sigs.k8s.io/clusterctl/commands/generate-cluster.html

File: docs/book/src/clusterctl/provider-contract.md
Comment (sbueringer) [reply docs/book/src/clusterctl/provider-contract.md:269]: Maybe, yup. But not if it is set in the refs in the ClusterClass.

Comment (sbueringer) [reply docs/book/src/clusterctl/provider-contract.md:269]: Given that we don't want to recommend using variables in ClusterClass we should probably recommend that the namespace should not be set at all.

So that then the combination of fixTargetNamespace and our defaulting webhook will make sure the right namespace is used. Otherwise things can get messy if the top-level namespace is overwritten and is not the same as the one hard-coded in the refs.

File: docs/book/src/clusterctl/provider-contract.md
Comment (enxebre) [reply docs/book/src/clusterctl/provider-contract.md:271]: If this something we want to prevent from happening and validate with webhooks?

Comment (fabriziopandini) [reply docs/book/src/clusterctl/provider-contract.md:271]: > If this something we want to prevent from happening and validate with webhooks?

Advanced users can choose to opt-in and reusing templates, it could be a powerful feature in the proper hands.
However, but given that in this context are talking about template for the quick start UX let reccomand the safest path.

File: docs/book/src/clusterctl/provider-contract.md
Comment (ykakarap) [reply docs/book/src/clusterctl/provider-contract.md:269]: >What would happen if someone sets a specific namespace? Would it be overwritten automatically or only when ${NAMESPACE} is used?

It will be overridden automatically. 

As long as we have a note saying that none of the references in the ClusterClass set the namespace things should work as expected. (Will add this note in the contract.)

>Given that we don't want to recommend using variables in ClusterClass we should probably recommend that the namespace should not be set at all.

We can definitely add this recommendation to make things consistent. But just FYI, it is totally fine if a namespace is specified in the resource. As long as none of the references use a namespace it will work.

File: docs/book/src/clusterctl/provider-contract.md
Comment (ykakarap) [reply docs/book/src/clusterctl/provider-contract.md:271]: >If this something we want to prevent from happening and validate with webhooks?

We would not want to prevent this as a feature in ClusterClass. As @fabriziopandini mentioned reusing templates can become a really powerful feature that advanced users will want to use. 

Given that this doc mostly focuses on 'example' ClusterClass definitions that a provider can provide, which in most cases will be used for quick start, we would want to recommend the safest path.

File: docs/book/src/clusterctl/provider-contract.md
Comment (ykakarap) [reply docs/book/src/clusterctl/provider-contract.md:273]: I dont see how this is in contrast with L263. L263 just says that the providers should document the list of available ClusterClass definitions.

File: docs/book/src/clusterctl/provider-contract.md
Comment (ykakarap) [reply docs/book/src/clusterctl/provider-contract.md:263]: I have added an additional note along these lines under the Variables section. Seemed more appropriate to add it under that section.

File: docs/book/src/clusterctl/provider-contract.md
Comment (sbueringer) [reply docs/book/src/clusterctl/provider-contract.md]: Got it, thx

File: docs/book/src/clusterctl/provider-contract.md
Comment (sbueringer) [reply docs/book/src/clusterctl/provider-contract.md:285]:
```suggestion
Currently the ClusterClass definitions SHOULD NOT have any variables in them.

ClusterClass definitions files should not use variable substitution, given that ClusterClass and managed topologies provide an alternative model for variable definition. 
```

**Review by fabriziopandini — COMMENTED** [review]
lgtm pending addressing open comments

File: docs/book/src/clusterctl/provider-contract.md
Comment (ykakarap) [reply docs/book/src/clusterctl/provider-contract.md:285]: I read it as the first sentence mentioning that ClusterClasses should not have variables (envsubst style variables as used in workload cluster templates) in them and the second sentence explains why we should not use variables in ClusterClass definitions. 

>Not sure if it is intended, but this sounds like the ClusterClasses should also not use variables (the new ones) as it writes that variables and variable substitutions should not be used.

By "the new ones" I am assuming you mean the variables used by for patching. If so, the doc is not talking about them.
How about "Currently the ClusterClass definitions SHOULD NOT have any environment variables in them." to avoid ambiguity.?

Comment (sbueringer) [reply docs/book/src/clusterctl/provider-contract.md]: Sounds great! My problem was only that the there's no way to tell with the current wording if the variables in the first sentence are ClusterClass variables.

**Comment by ykakarap** [comment]
The error in the "Broken Links" pipeline seems to be 
>ERROR: 1 dead links found!
[✖] https://github.com/kubernetes-sigs/cluster-api/issues/833 → Status: 0

This link is in the "Cluster Provider Spec and Status as CRDs" proposal doc and I believe the error is not related to changes in the current PR.

**Comment by fabriziopandini** [comment]
If I'm not wrong the link is correct, might be a flake...

**Comment by sbueringer** [comment]
/lgtm
Yeah looks correct. You can try a re-run by commit --amend + push --force

**Comment by ykakarap** [comment]
>Yeah looks correct. You can try a re-run by commit --amend + push --force

I wonder if this will remove the lgtm label. 
I believe someone with rights can just re-trigger the github workflow test form the UI.

**Comment by sbueringer** [comment]
> > Yeah looks correct. You can try a re-run by commit --amend + push --force
> 
> I wonder if this will remove the lgtm label. I believe someone with rights can just re-trigger the github workflow test form the UI.

Yup, it will remove them. Just suggested it because getting lgtm again shouldn't be a problem :)

I think someone with write or admin access to the repo could be able to restart the workflow, I'm not sure though.

**Comment by ykakarap** [comment]
@sbueringer can you please LGTM again 🙏🏼  
Hopefull the test doesn't flake this time :)

**Comment by sbueringer** [comment]
Sure :)

/lgtm

**Review by killianmuldoon — APPROVED** [review]
/lgtm

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

@ykakarap please give a PSA in the office hours about this documentation being available

**Pull request merged**
